### ISSUE TITLE
[Proposal] MediaElement

### ISSUE THREAD

**Issue by brminnick** [CommunityToolkit/Maui#113]
# MediaElement

* [x] Proposed
* [x] Prototype: Not Started
* [x] Implementation: Not Started
  * [x] iOS Support
  * [x] Android Support
  * [x] macOS Support
  * [x] Windows Support
* [x] Unit Tests: Not Started
* [x] Sample: Not Started
* [x] Documentation: Not Started

## Summary
[summary]: #summary

MediaElement is a view for playing video and audio. Media that's supported by the underlying platform can be played from the following sources:
- The web, using a URI (HTTP or HTTPS).
- A resource embedded in the platform application, using the ms-appx:/// URI scheme.
- Files that come from the app's local and temporary data folders, using the ms-appdata:/// URI scheme.
The device's library.

## Detailed Design
[design]: #detailed-design

### MediaElement.shared.cs

```cs
public class MediaElement : View, IMediaElementController
{
  public static readonly BindableProperty AspectProperty;
  public static readonly BindableProperty AutoPlayProperty;
  public static readonly BindableProperty BufferingProgressProperty;
  public static readonly BindableProperty CurrentStateProperty;
  public static readonly BindableProperty DurationProperty;
  public static readonly BindableProperty IsLoopingProperty;
  public static readonly BindableProperty KeepScreenOnProperty;
  public static readonly BindableProperty PositionProperty;
  public static readonly BindableProperty ShowsPlaybackControlsProperty;
  public static readonly BindableProperty SourceProperty;
  public static readonly BindableProperty VideoHeightProperty;
  public static readonly BindableProperty VideoWidthProperty;
  public static readonly BindableProperty VolumeProperty;
  public static readonly BindableProperty SpeedProperty;
  
  public double BufferingProgress { get; }
  public bool CanSeek { get; }
  public MediaElementState CurrentState { get; }
  public TimeSpan? Duration { get; }
  public int VideoHeight { get; }
  public int VideoWidth  { get;}
  
  public Aspect Aspect { get; set; }
  public bool AutoPlay { get; set; }
  public bool IsLooping { get; set; }
  public bool KeepScreenOn { get; set; }
  public bool ShowsPlaybackControls { get; set; }
  public TimeSpan Position { get; set; }
  [Forms.TypeConverter(typeof(MediaSourceConverter))]
  public MediaSource? Source { get; set; }
  public double Volume { get; set; }
  public double Speed { get; set; }
  
  public event EventHandler? MediaEnded;
  public event EventHandler? MediaFailed;
  public event EventHandler? MediaOpened;
  public event EventHandler? SeekCompleted;
  
  public void Play() ;
  public void Pause();
  public void Stop();
}
```

## Usage Syntax
[usage]: #usage-syntax

<!-- Here is an example from CharactersValidationBehavior:

### XAML Usage

```xml
<MediaElement Source="https://sec.ch9.ms/ch9/5d93/a1eab4bf-3288-4faf-81c4-294402a85d93/XamarinShow_mid.mp4"
              ShowsPlaybackControls="True" />
```

### C# Usage

```cs
new MediaElement
{
  Source = "https://sec.ch9.ms/ch9/5d93/a1eab4bf-3288-4faf-81c4-294402a85d93/XamarinShow_mid.mp4"
}
```

**Comment by jfversluis** [comment]
The `MediaElement` needs some love, I think what is there now is pretty stable, lots of room for improvement though. But I think a control to play media should actually be part of .NET MAUI in-the-box. While it's not that, I think we should definitely have it in the Toolkit to provide something, hopefully build it out and move it over to .NET MAUI as a mature control.

**Comment by pierre01** [comment]
MediaElement is essential to many applications and should be added to the Maui.Essentials library the same as haptic feedback is.

**Comment by jfversluis** [comment]
The big difference is that Essentials does not have any UI elements. It will be implemented here as it is in the Toolkit now and hopefully we can transition it into .NET MAUI itself.

**Comment by peterfoot** [comment]
I had thought about whether it could be ported to MAUI and if so what changes, if any, could be made. One thing which did occur - should it follow the UWP model which has seen the MediaElement deprecated for a MediaPlayer and MediaPlayerElement. The latter being the UI control and the former having all the mechanics to playback media. It would mean more flexibility for audio only and not needing to add a UI control in those cases. I think the underlying platform APIs in iOS and Android would still suit this model...

**Comment by peterfoot** [comment]
And I agree with @jfversluis it would be awesome for this control to transition into MAUI itself.

**Comment by jfversluis** [comment]
@peterfoot if you're up for it, it's yours for the taking here :)

I think timing-wise we won't be able to get this into .NET MAUI straight-away, so let's get it in here first and then we can start making a case to move it for .NET 7.

**Comment by emceelovin** [comment]
A/V is paramount IMO. This should've been on the timeline from day one *facepalm*. I'm all for whatever will allow basic playback, but the proposal doesn't mention anything about a Stream as one the source options, and that's the most important IMO. I'm working on a tiny IMediaPlayer interface in my MAUI project to back with native implementations, but maybe I can help on this? I would like to see this ASAP also.

**Comment by brminnick** [comment]
Thanks @emceelovin! We would love your help with MediaElement 💯

**Comment by emceelovin** [comment]
@brminnick will do anything I can!

**Comment by brminnick** [comment]
Adding the `blocked` label for now.

We agree that `MediaElement` should support streaming. For Android, supporting streaming requires the use of Google's Exo Player, an [open-source library made by Google](https://github.com/google/ExoPlayer) that enables `HLS` and `DASH` on Android.

Currently, there is no `net6.0-android` binding of Google's ExoPlayer. This is blocking us from adding `MediaElement`.

I've spoken with David Ortinau and Jon Dick on the .NET MAUI Engineering team about this gap in our feature offering. They are currently evaluating the feasibility of creating an official Microsoft-created ExoPlayer binding for `net6.0-android`. 

For Xamarin.Forms developers, the current recommendation is to use the community-created [`Xam.Plugins.Android.ExoPlayer` NuGet Package](https://www.nuget.org/packages/Xam.Plugins.Android.ExoPlayer/) by Martijn van Dijk. I use Martijn's ExoPlayer NuGet package in my app that I've published to the App Stores, [GitTrends](https://github.com/brminnick/GitTrends/blob/main/GitTrends.Android/CustomRenderers/VideoPlayerViewCustomRenderer.cs), and it works really well to stream video from [Azure Media Services](https://azure.microsoft.com/services/media-services/?WT.mc_id=mobile-0000-bramin)! Here's the ExoPlayer source code in GitTrends for those curious: https://github.com/brminnick/GitTrends/blob/main/GitTrends.Android/CustomRenderers/VideoPlayerViewCustomRenderer.cs

**Comment by VladislavAntonyuk** [comment]
https://github.com/dotnet/maui/issues/6764

**Comment by sturlath** [comment]
Please also consider DRM (with/without token) ans AES token encryption for streaming.

**Comment by jabak** [comment]
> Please also consider DRM (with/without token) ans AES token encryption for streaming.

I agree. I would consider this to be part of basic audio playback. 

Playing/looping audio from a memstream (c# memstream etc.), an url (https) or a file (beep.wav - beep.mp3) including DRM content is what is needed for common cross platform apps.

**Comment by jabak** [comment]
> @sturlath and @jabak do you by any chance have information or links to information that helps to understand what is needed to support those things?

You mean what libs etc. on the various platforms that could be used to support this?

**Comment by bijington** [comment]
> > @sturlath and @jabak do you by any chance have information or links to information that helps to understand what is needed to support those things?
> 
> 
> 
> You mean what libs etc. on the various platforms that could be used to support this?
> 
> 

Well anything that may assist at the implementation stage. I suspect DRM support might not be straightforward (although that is an assumption).

We are going to be limited to what the platforms provide, we try our best to not bring in any external dependencies. Of course if it was essential perhaps the Media part of the toolkit could become its own package.

**Comment by jabak** [comment]
I think DRM support would be a challenge without taking external dependencies.

Also I think audio playback at a very basic level is essential. Accessibility comes into mind first and foremost. So being able to play simple sounds from memory streams (wav format etc.) and small audio files is a must. 

Url-based streams and DRM is next level in my opinion.  But still basic audio playback nevertheless.

I've have no clue what the various platforms provide in terms of audio playback these day. In the past I've used the NAudio library for .NET Console Apps and UWP/WinUI apps. But that's not a built-in thingy. The Windows Media Player is not an option - I think it's an ActiveX component. For Xamarin I've played around with Plugin.MediaManager - that worked okay actually.

Sorry I can't point you to better options.

**Comment by bijington** [comment]
No need to apologise, it is discussions like this that help us on the journey to finding a more complete solution. Thanks for the links

**Comment by charlesroddie** [comment]
Currently MediaElement from Xamarin.CommunityToolkit doesn't work on [Windows](https://github.com/xamarin/XamarinCommunityToolkit/issues/1874) or [Android](https://github.com/xamarin/XamarinCommunityToolkit/issues/1650)!

https://github.com/inthehand/InTheHand.Forms works so may be a better source for starting this.

**Comment by GuildOfCalamity** [comment]
It's strange there's no audio capability for MAUI, since this was already in the Xamarin framework.

**Comment by bijington** [comment]
> It's strange there's no audio capability for MAUI, since this was already in the Xamarin framework.

It's funny you should mention it because this was being worked on https://github.com/jfversluis/Plugin.Maui.Audio 👀

**Comment by JohnGalt1717** [comment]
Re DRM:

DRM on windows is trivial. The MediaElement there has everything in a few lines of code to support both AES encryption and play ready.

Android is similarly easy using Exo player and google's stuff which is built in.
iOS is the problem child and requires a VERY different approach to building up the playback support to enable AES and Fairplay. It needs to be done right from the beginning or you'll be doing it again to support other stuff too (i.e. captioning from external files will also require this approach.

Also as a side note, iOS now requires the use of their player and NO OTHER players are allowed. That is, you can't use your own platform layer to do it, so this has to be native all of the way through and done their way. I'm not sure why there isn't an Apple provided wrapper that has everything baked in given their new requirements.

**Comment by brminnick** [comment]
Removing the `blocked` label now that [Xam.Plugins.Android.ExoPlayer](https://www.nuget.org/packages/Xam.Plugins.Android.ExoPlayer) has added support for `net6.0-android`

**Comment by mhrastegari** [comment]
Hi @jfversluis Love your work, the MCT MediaElement saved my day but i have a problem with android version! everything works fine on Native controls but when i use custom controls for it not working well.

when i toggle play and pause like this
`if (media.CurrentState == MediaElementState.Playing)
            media.Pause();
        else if (media.CurrentState == MediaElementState.Paused || media.CurrentState == MediaElementState.Stopped)
            media.Play();`
i cannot play the video again, so i should seek it backward and forward multiple time to make it work.

**Comment by pictos** [comment]
@mhrastegari can you share a link that reproduces the issue?
Can you check the value of the `Playing` property when you pause the video?

**Comment by mhrastegari** [comment]
@pictos  there is no `Playing` property for MediaElement, there is only a enum of it. 🥲
funny thing i put a BreakPoint on top of the method and app crashed!

I'll share a sample repo ASAP.

**Comment by mhrastegari** [comment]
@jfversluis @pictos Here is a full sample.
[MediaElementTest.zip](https://github.com/CommunityToolkit/Maui/files/9936786/MediaElementTest.zip)

FYI : the test media sources are defined here
![image](https://user-images.githubusercontent.com/42671084/199937661-c782c596-9f16-4cb1-9d72-dc85bc719066.png)

**Comment by jfversluis** [comment]
Hey @mhrastegari that's good feedback, exactly what we are looking for, thanks!

For Android you really have to call a method to prepare the media for playback, I wonder if that is something that we should do if you call play again and if that is why it's not working... I've added this finding to my little todo list in #717

**Comment by mhrastegari** [comment]
@jfversluis  Not sure this helps but i put a breakpoint on my Play/Pause method and my debugger stucks here in a loop.
any idea or workarounds? I really have to implement this in our App for tomorrow. 🥲

![image](https://user-images.githubusercontent.com/42671084/199967475-e6800e8b-0ccc-4c7a-83a8-a8d893b4f930.png)

**Comment by jfversluis** [comment]
@mhrastegari I'm not trying to be rude or mean this the wrong way, but you're trying to use code that hasn't been released and is at best in alpha stage. I appreciate your enthusiasm for wanting to try this out and use this in an app, but at this point we can't really guarantee this level of support. For what it's worth, this is a free, community maintained library. Even if it's released still can't guarantee this level of support to fix things within 24 hours. Just wanted to make that clear.

Normally, I would still do my best to try and help you out. Unfortunately, I don't have time today to look into this. If you do want to be unblocked you'll either have to find a solution yourself or maybe just take the code that I have implemented and add that directly to your project and see what is not working for you.

Hope you can get it done and make something amazing! Good luck!

**Comment by mhrastegari** [comment]
Yeah you are right :( I'm really sad right now about how our conversation got there, I was just in too much stress and i should have think more about what i going to comment here. 
I'm so sorry dear Gerald you're amazing. ❤️

**Comment by jfversluis** [comment]
Don't worry! No harm was done, we're still friends and I understand it comes from a good place. But also want to make sure that you have expectations we can't live up to.

Thanks for saying this though, much appreciated!

**Comment by Flaechekueste50** [comment]
Hey! I just tried out the MediaElement in my project (an audio streaming radio app) and I've got following Exception which I really don't understand at all:

![image](https://user-images.githubusercontent.com/88831296/200844329-3067f47c-0218-4283-a5f8-bc46774ae950.png)

Are you able to tell me what this Exception means and how I can solve the problem?

**Comment by jfversluis** [comment]
@Flaechekueste50 as above, not really. It's being actively developed, no support is given at this time. If you can give me a source URL that will stay online for a while I can add it to #717 and test it out to see what we can do to make it work.

**Comment by aawaheed** [comment]
Will there be any support for streaming over UDP?

**Comment by mhrastegari** [comment]
Hey guys these 3 days we were discovered a serious memory leak in MCT MediaElement in our project and we fixed it in Android and Windows(not tested on macOS and iOS), and just wanna return the favor, specially to @jfversluis :) 
So how can i help and contribute?

**Comment by pictos** [comment]
@mhrastegari thanks for that <3

First you should work against this [branch](https://github.com/CommunityToolkit/Maui/tree/add-mediaelement), when you've everything working just open a PR targeting this branch

**Comment by mhrastegari** [comment]
@pictos Done.

**Comment by jfversluis** [comment]
Appreciate it @mhrastegari thank you so much!

**Comment by PaulVrugt** [comment]
Is my assumption correct that this will support a Stream as source? If so, this will finally fix the issue that we have had for ages that we need to be able to pass a custom header to a uri, since the video files we are playing are coming from an api that needs authentication via the `authorization` header.

**Comment by jfversluis** [comment]
@PaulVrugt not sure if that will be possible with the first version. What would passing headers through a stream look like? Do you have some pseudo code or a more thorough explanation of what you want to do?

**Comment by vynxcc** [comment]
> @PaulVrugt not sure if that will be possible with the first version. What would passing headers through a stream look like? Do you have some pseudo code or a more thorough explanation of what you want to do?

![image](https://user-images.githubusercontent.com/93080410/208045654-e0a3f12a-61f7-465e-a2a9-55e674bb9fcb.png)

**Comment by vynxcc** [comment]
> > @PaulVrugt not sure if that will be possible with the first version. What would passing headers through a stream look like? Do you have some pseudo code or a more thorough explanation of what you want to do?
> 
> ![image](https://user-images.githubusercontent.com/93080410/208045654-e0a3f12a-61f7-465e-a2a9-55e674bb9fcb.png)

this is hard coded but im sure you get the idea 😁

**Comment by ne0rrmatrix** [comment]
I have been testing against the nuget package. It has been a great learning experience for me. This is fun! I am working on developing an app to play podcasts from a set of rss feeds from twit.tv. 

It is working in Windows and Android. Still a lot for me to do to have it ready for testing though. I am planning on looking into helping out here and fixing the issues I am having with maui. I am new to developing and have never contributed to an open source project. I am hoping to make this my first. 

I am going to go and work on  fix to a very minor issue with media element. It might take a week or so to see if I have any hope to help with this. The first thing I want to fix is a problem where if you don't specify in xaml a WidthRequest and HeightRequest that fits within the android screen the ShowPlayBackControls is missing. Also the video does not show. If you do specify the width and height it works. Has horribly stuttering with the video I am playing. It is streaming over the internet. Not going to try and fix that first lol. I am going to work on something small like the UI issue. Wish me luck!

**Comment by mhrastegari** [comment]
Hey @ne0rrmatrix i believe this can help you for your fullscreen issue.

https://github.com/davidbritch/dotnet-maui-videoplayer/issues/9#issuecomment-1323488802

**Comment by JoacimWall** [comment]
Hi try to find the source code for the Media Element to test I I can add support for subtiles on Android but can't find the source. Is not the MediaElement open source?

//Thanks Joacim

**Comment by bijington** [comment]
> Hi try to find the source code for the Media Element to test I I can add support for subtiles on Android but can't find the source. Is not the MediaElement open source?
> 
> 
> 
> //Thanks Joacim  

The code hasn't been merged into main yet. The work is still ongoing under the PR https://github.com/CommunityToolkit/Maui/pull/667

**Comment by JoacimWall** [comment]
> > Hi try to find the source code for the Media Element to test I I can add support for subtiles on Android but can't find the source. Is not the MediaElement open source?
> > //Thanks Joacim
> 
> The code hasn't been merged into main yet. The work is still ongoing under the PR #667

Thanks 
Next question, Now when I have fixed the support for subtiles on android how do I submit this. I can't not figure out if there is any way to create a Pull Request on an already existing PR(667). 

//Thanks Joacim

**Comment by jfversluis** [comment]
@JoacimWall let's wait until `MediaPlayer` has been merged. That PR is way too big already, I don't want to delay it any further by doing a PR against the PR.

If you really need it right now then you should be able to indeed do it with a custom mapper entry. Something like this:

```csharp
MediaPlayerHandler.PropertyMapper.AppendToMapping("subtitleMapper", (handler, player) =>
		{
#if ANDROID
			handler.PlatformView.GetFirstChildOfType<StyledPlayerView>()?.SetShowSubtitleButton(true);
#endif
		});
```

**Comment by JoacimWall** [comment]
> @JoacimWall let's wait until `MediaPlayer` has been merged. That PR is way too big already, I don't want to delay it any further by doing a PR against the PR.
> 
> If you really need it right now then you should be able to indeed do it with a custom mapper entry. Something like this:
> 
> ```cs
> MediaPlayerHandler.PropertyMapper.AppendToMapping("subtitleMapper", (handler, player) =>
> {
> #if ANDROID
>     handler.PlatformView.GetFirstChildOfType<StyledPlayerView>()?.SetShowSubtitleButton(true);
> #endif
> });
> ```

Ok no problem. Thanks for the handler code. 
//Joacim

**Comment by ne0rrmatrix** [comment]
With stuff still in early alpha on this build what sort of things can we do that will be helpful? I am learning about maui, mvvm as quick as I can. I am going through and learning about maui and trying to figure out all the ins and outs so I can contribute by fixing bugs myself. I want to help and provide solutions to issues not just submit bugs. I have all the time in the world and want to contribute.  I have learned a lot already. But I am in the process of trying to figure out how to understand how the community toolkit and this project fit together with dotnet maui. I plan to keep figuring it out over the next month.

I have learned how to use the community toolkit to do almost all of things I want. But I need to learn the maui long hand way to contribute and make my life easier I think? Most of the issues I have had with maui so far I have found to be either user errors on my part on how I implement something. I am this week going to spend my time going over and doing everything the long way so I can learn the structure of how maui is done without the toolkit. This is in the hopes of making it easier to learn how to contribute to toolkit and understand the structure of the code. I have been going through the toolkit and looking at the code.
